File: backend/routes/birth_cards.py
"""
Birth-card endpoints (Greer "Lifetime Cards"). Pure computation over a
birth date via the root birth_cards module — nothing is stored except
the two display preferences (method, 8/11 naming) in settings.

Card references come back hydrated with display names and, where a
match exists, the Tarot card_archetypes id so the UI can link into the
Reference section.
"""

from datetime import date

from flask import Blueprint, jsonify, request, current_app

import birth_cards as bc
from backend.utils import require_json

birth_cards_bp = Blueprint('birth_cards', __name__)

METHOD_KEY = 'birth_cards_method'
EIGHT_ELEVEN_KEY = 'birth_cards_eight_eleven'
COURT_SYSTEM_KEY = 'birth_cards_court_system'


def _prefs(db):
    method = db.get_setting(METHOD_KEY) or bc.GREER
    eight_eleven = db.get_setting(EIGHT_ELEVEN_KEY) or 'golden_dawn'
    court_system = db.get_setting(COURT_SYSTEM_KEY) or 'golden_dawn'
    if method not in bc.METHODS:
        method = bc.GREER
    if eight_eleven not in ('golden_dawn', 'marseille'):
        eight_eleven = 'golden_dawn'
    if court_system not in bc.COURT_SYSTEMS:
        court_system = 'golden_dawn'
    return method, eight_eleven, court_system


def tarot_archetype_ids(db):
    """Map (rank, suit) -> archetype id for every Tarot archetype.
    Majors use rank '0'..'21' with suit 'Major Arcana'; Minors are
    matched by name instead (rank strings differ per deck tradition)."""
    cursor = db.conn.cursor()
    cursor.execute(
        "SELECT id, name, rank, suit FROM card_archetypes "
        "WHERE cartomancy_type = 'Tarot'")
    by_rank = {}
    by_name = {}
    for row in cursor.fetchall():
        r = row if isinstance(row, dict) else dict(row)
        if r['suit'] == 'Major Arcana' and r['rank'] is not None:
            by_rank[str(r['rank'])] = r['id']
        by_name[r['name'].lower()] = r['id']
    return by_rank, by_name


def default_tarot_card_ids(db):
    """Map archetype name (lowercased) -> card id in the user's default
    Tarot deck, so the UI can show real card images. Empty if no
    default deck is set."""
    deck_id = db.get_default_deck('Tarot')
    if not deck_id:
        return {}
    cursor = db.conn.cursor()
    cursor.execute(
        'SELECT id, name, archetype FROM cards WHERE deck_id = ? '
        'ORDER BY card_order', (deck_id,))
    out = {}
    for row in cursor.fetchall():
        r = row if isinstance(row, dict) else dict(row)
        for key in (r.get('archetype'), r.get('name')):
            if key and key.lower() not in out:
                out[key.lower()] = r['id']
    return out


def _hydrate(profile, eight_eleven, court_system, db):
    """Attach display names, archetype ids, and (when a default Tarot
    deck is set) card ids for images to every card reference."""
    by_rank, by_name = tarot_archetype_ids(db)
    card_ids = default_tarot_card_ids(db)

    def major(n):
        if n is None:
            return None
        # Image lookup tries both 8/11 labels plus Thoth aliases, so a
        # Marseille-style or Thoth default deck still resolves.
        candidates = [bc.major_name(n, eight_eleven), bc.MAJOR_NAMES[n]]
        if n == 8:
            candidates.append('Justice')
        if n == 11:
            candidates.append('Strength')
        candidates.extend(bc.MAJOR_ALIASES.get(n, []))
        card_id = next(
            (card_ids[c.lower()] for c in candidates if c.lower() in card_ids),
            None)
        return {
            'number': n,
            'name': bc.major_name(n, eight_eleven),
            'archetype_id': by_rank.get(bc.major_archetype_rank(n)),
            'card_id': card_id,
        }

    def minor(ref):
        name = bc.card_ref_name(ref)
        return {
            'rank': ref['rank'],
            'suit': ref['suit'],
            'name': name,
            'archetype_id': by_name.get(name.lower()),
            'card_id': card_ids.get(name.lower()),
        }

    def ruler_major(n):
        """Like major(), but always the canonical Golden Dawn name:
        astrological attributions are card identities (Leo IS the
        Strength card), so the 8/11 numbering toggle doesn't apply."""
        hydrated = major(n)
        hydrated['name'] = bc.MAJOR_NAMES[n]
        return hydrated

    rulers = bc.zodiacal_rulers(profile['zodiacal_card'])
    court = bc.decan_court(profile['zodiacal_card'], court_system)

    out = dict(profile)
    out['zodiacal_rulers'] = rulers
    out['decan_court'] = court
    out['cards'] = {
        'personality': major(profile['personality']),
        'soul': major(profile['soul']),
        'teacher': major(profile['teacher']),
        'hidden_factor': [major(n) for n in profile['hidden_factor']],
        'constellation_majors': [major(n) for n in profile['constellation']['majors']],
        'lessons_and_opportunities': [
            minor(ref) for ref in profile['lessons_and_opportunities']],
        'zodiacal': minor(profile['zodiacal_card']),
        'zodiacal_sign_ruler': ruler_major(rulers['sign_major']),
        'zodiacal_planet_ruler': ruler_major(rulers['planet_major']),
        'decan_court': {
            'rank': court['rank'],
            'suit': court['suit'],
            'name': court['name'],
            'archetype_id': by_name.get(court['name'].lower()),
            'card_id': card_ids.get(court['name'].lower()),
        },
    }
    for key in ('year_card', 'generic_year', 'personal_month'):
        if key in profile:
            out['cards'][key] = major(profile[key])
    return out


def _compute_response(db, birth: date, birth_date_str: str):
    method = request.args.get('method')
    eight_eleven = request.args.get('eight_eleven')
    court_system = request.args.get('court_system')
    saved_method, saved_ee, saved_court = _prefs(db)
    method = method if method in bc.METHODS else saved_method
    eight_eleven = eight_eleven if eight_eleven in ('golden_dawn', 'marseille') else saved_ee
    court_system = court_system if court_system in bc.COURT_SYSTEMS else saved_court

    today = date.today()
    try:
        reference_year = int(request.args.get('year', today.year))
        reference_month = int(request.args.get('month', today.month))
    except ValueError:
        return jsonify({'error': 'year and month must be integers'}), 400

    profile = bc.calculate(
        birth, method=method,
        reference_year=reference_year, reference_month=reference_month)
    hydrated = _hydrate(profile, eight_eleven, court_system, db)

    age = today.year - birth.year - (
        (today.month, today.day) < (birth.month, birth.day))
    hydrated['birth_date'] = birth_date_str
    hydrated['age'] = age
    hydrated['eight_eleven'] = eight_eleven
    hydrated['court_system'] = court_system
    hydrated['reference_year'] = reference_year
    hydrated['reference_month'] = reference_month
    return jsonify(hydrated)


@birth_cards_bp.route('/api/profiles/<int:profile_id>/birth-cards')
def profile_birth_cards(profile_id):
    db = current_app.config['DB']
    row = db.get_profile(profile_id)
    if not row:
        return jsonify({'error': 'Profile not found'}), 404
    profile = row if isinstance(row, dict) else dict(row)
    birth_str = profile.get('birth_date')
    if not birth_str:
        return jsonify({'error': 'Profile has no birth date'}), 400
    try:
        birth = date.fromisoformat(birth_str)
    except ValueError:
        return jsonify({'error': f'Unreadable birth date: {birth_str}'}), 400
    return _compute_response(db, birth, birth_str)


@birth_cards_bp.route('/api/birth-cards')
def adhoc_birth_cards():
    """Birth cards for an arbitrary date (no profile needed)."""
    db = current_app.config['DB']
    date_str = request.args.get('date', '')
    try:
        birth = date.fromisoformat(date_str)
    except ValueError:
        return jsonify({'error': 'date must be YYYY-MM-DD'}), 400
    return _compute_response(db, birth, date_str)


@birth_cards_bp.route('/api/birth-cards/prefs')
def get_birth_card_prefs():
    method, eight_eleven, court_system = _prefs(current_app.config['DB'])
    return jsonify({'method': method, 'eight_eleven': eight_eleven,
                    'court_system': court_system})


@birth_cards_bp.route('/api/birth-cards/prefs', methods=['PUT'])
@require_json
def set_birth_card_prefs(data):
    db = current_app.config['DB']
    method = data.get('method')
    eight_eleven = data.get('eight_eleven')
    court_system = data.get('court_system')
    if method is not None:
        if method not in bc.METHODS:
            return jsonify({'error': f'method must be one of {bc.METHODS}'}), 400
        db.set_setting(METHOD_KEY, method)
    if eight_eleven is not None:
        if eight_eleven not in ('golden_dawn', 'marseille'):
            return jsonify({'error': 'eight_eleven must be golden_dawn or marseille'}), 400
        db.set_setting(EIGHT_ELEVEN_KEY, eight_eleven)
    if court_system is not None:
        if court_system not in bc.COURT_SYSTEMS:
            return jsonify({'error': f'court_system must be one of {bc.COURT_SYSTEMS}'}), 400
        db.set_setting(COURT_SYSTEM_KEY, court_system)
    method, eight_eleven, court_system = _prefs(db)
    return jsonify({'method': method, 'eight_eleven': eight_eleven,
                    'court_system': court_system})

File: frontend/src/api/birthCards.ts
import api from './client';

// === Birth cards (Greer "Lifetime Cards") ===

export type BirthCardMethod = 'greer' | 'amberstone';
export type EightEleven = 'golden_dawn' | 'marseille';
export type CourtSystem = 'golden_dawn' | 'golden_dawn_waite' | 'bota';

/** A Major Arcana reference, hydrated by the backend. */
export interface MajorCardRef {
  /** 1–22; The Fool is 22 internally (displayed as 0 by convention). */
  number: number;
  name: string;
  archetype_id: number | null;
  /** Matching card in the default Tarot deck, for the image. */
  card_id: number | null;
}

/** A Minor Arcana reference (decan / lessons cards). */
export interface MinorCardRef {
  rank: number;
  suit: string;
  name: string;
  archetype_id: number | null;
  card_id: number | null;
}

export interface BirthCardProfile {
  method: BirthCardMethod;
  base_number: number;
  personality: number;
  soul: number;
  teacher: number | null;
  hidden_factor: number[];
  pattern: string;
  nighttime: boolean;
  fool_center: boolean;
  constellation: { root: number; majors: number[] };
  dynamic: 1 | 2 | 3 | null;
  /** Golden Dawn decan attributions for the zodiacal card. */
  zodiacal_rulers: {
    sign: string;
    planet: string;
    sign_major: number;
    planet_major: number;
  };
  /** Court card ruling the birth decan's 20°–20° span. */
  decan_court: {
    rank: string;
    suit: string;
    name: string;
    court_sign: string;
    span: string;
  };
  court_system: CourtSystem;
  /** Year Card per year, birth year through ~10 years ahead. */
  year_series: { year: number; card: number }[];
  majors_by_number: Record<string, MajorCardRef>;
  karmic_year: number;
  year_card: number;
  generic_year: number;
  personal_month: number;
  birth_date: string;
  age: number;
  eight_eleven: EightEleven;
  reference_year: number;
  reference_month: number;
  cards: {
    personality: MajorCardRef;
    soul: MajorCardRef;
    teacher: MajorCardRef | null;
    hidden_factor: MajorCardRef[];
    constellation_majors: MajorCardRef[];
    lessons_and_opportunities: MinorCardRef[];
    zodiacal: MinorCardRef;
    zodiacal_sign_ruler: MajorCardRef;
    zodiacal_planet_ruler: MajorCardRef;
    decan_court: {
      rank: string;
      suit: string;
      name: string;
      archetype_id: number | null;
      card_id: number | null;
    };
    year_card: MajorCardRef;
    generic_year: MajorCardRef;
    personal_month: MajorCardRef;
  };
}

export interface BirthCardPrefs {
  method: BirthCardMethod;
  eight_eleven: EightEleven;
  court_system: CourtSystem;
}

interface BirthCardOpts {
  method?: BirthCardMethod;
  eightEleven?: EightEleven;
  courtSystem?: CourtSystem;
  /** Reference year for the Year Card (defaults server-side to today). */
  year?: number;
}

function optsToParams(opts?: BirthCardOpts): URLSearchParams {
  const params = new URLSearchParams();
  if (opts?.method) params.set('method', opts.method);
  if (opts?.eightEleven) params.set('eight_eleven', opts.eightEleven);
  if (opts?.courtSystem) params.set('court_system', opts.courtSystem);
  if (opts?.year) params.set('year', String(opts.year));
  return params;
}

export async function getProfileBirthCards(
  profileId: number,
  opts?: BirthCardOpts,
): Promise<BirthCardProfile> {
  const qs = optsToParams(opts).toString();
  const res = await api.get(`/api/profiles/${profileId}/birth-cards${qs ? `?${qs}` : ''}`);
  return res.data;
}

export async function getBirthCardsForDate(
  date: string,
  opts?: BirthCardOpts,
): Promise<BirthCardProfile> {
  const params = optsToParams(opts);
  params.set('date', date);
  const res = await api.get(`/api/birth-cards?${params.toString()}`);
  return res.data;
}

export async function getBirthCardPrefs(): Promise<BirthCardPrefs> {
  const res = await api.get('/api/birth-cards/prefs');
  return res.data;
}

export async function setBirthCardPrefs(
  prefs: Partial<BirthCardPrefs>,
): Promise<BirthCardPrefs> {
  const res = await api.put('/api/birth-cards/prefs', prefs);
  return res.data;
}
